<a href="https://colab.research.google.com/github/ayushhbroshan/Projects/blob/main/West_bengal_election_2026.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🗳️ Project Introduction: West Bengal Assembly Election 2026 – Candidate Analytics

This notebook presents a comprehensive **Exploratory Data Analysis (EDA)** of the **West Bengal Assembly Election 2026 Candidate Dataset**, focusing on the financial, educational, and criminal profiles of contesting candidates. Using Python and data visualization techniques, the analysis aims to uncover meaningful insights into candidate demographics, party-wise trends, constituency-level patterns, and financial disclosures.

The objective of this project is to transform raw election affidavit data into actionable insights that can support **political consulting, election research, media reporting, and data-driven decision making**.

<div align="center">
  <img src="https://images.unsplash.com/photo-1529107386315-e1a2ed48a620?auto=format&fit=crop&w=1400&q=80" width="100%" alt="Election Analytics Banner">
</div>

---

## 📌 Data Source

**Source:** MyNeta Election Watch

🔗 https://myneta.info/WestBengal2026/index.php?action=show_winners&sort=candidate

The dataset has been collected from **MyNeta**, a well-known public election information platform developed by the **Association for Democratic Reforms (ADR)** and the **National Election Watch (NEW)**. The platform compiles information from candidates' **official election affidavits** submitted to the Election Commission of India, making it a widely used and credible source for election-related research and analysis.

---

## 🎯 Project Objectives

- Analyze the financial strength of election candidates through declared **Assets** and **Liabilities**.
- Compare political parties based on candidate wealth, liabilities, education, and criminal records.
- Identify financially influential and high-risk candidates using **Debt Ratio Analysis**.
- Examine the educational background of candidates across parties.
- Explore constituency-wise candidate profiles and financial competitiveness.
- Build insightful visualizations for executive-level reporting and political intelligence.

---

## 📂 Dataset Overview

The dataset contains affidavit information of candidates contesting the **West Bengal Assembly Election 2026**.

| Feature | Description |
|----------|-------------|
| Sno | Candidate Serial Number |
| Candidate | Candidate Name |
| Constituency | Assembly Constituency |
| Party | Political Party |
| Criminal Cases | Number of Declared Criminal Cases |
| Education | Highest Educational Qualification |
| Assets_Amount | Total Declared Assets (₹) |
| Liabilities_Amount | Total Declared Liabilities (₹) |

---

## 📊 Key Areas of Analysis

- 💰 Candidate Wealth Analysis
- 📉 Liability & Debt Analysis
- 🏛 Party-wise Comparative Analysis
- 🚨 Criminal Background Analysis
- 🎓 Education Profile Analysis
- 🗺 Constituency Intelligence
- 📈 Financial & Political Insights
- 📊 Data Visualization Dashboard

---

## 🛠️ Technologies Used

- Python
- Pandas
- NumPy
- Matplotlib
- Google Colab

---

>**Goal:** Develop a comprehensive, data-driven analysis of candidates contesting the **West Bengal Assembly Election 2026**. Since elected candidates represent constituencies across the state, their financial, educational, and criminal profiles collectively provide valuable insights into West Bengal's political landscape, candidate quality, and socio-economic representation. The analysis aims to support political consulting, election research, and evidence-based decision making through meaningful visualizations and statistical insights.

In [ ]:
import pandas as pd
import numpy as np
df=pd.read_csv('https://raw.githubusercontent.com/ayushhbroshan/files/refs/heads/main/west_bengal_2026_complete_292.csv')
df

,Sno,Candidate,Constituency,Party,Criminal Cases,Education,Assets,Liabilities
0,1,Abdul Aziz Doctor,LALGOLA,AITC,0,Post Graduate,"Rs 6,29,08,242 ~ 6 Crore+","Rs 38,50,000 ~ 38 Lacs+"
1,2,Abdul Khaleque Molla,METIABURUZ,AITC,2,10th Pass,"Rs 1,74,55,505 ~ 1 Crore+",Rs 0 ~
2,3,Abdul Matin Muhammad,HAROA,AITC,1,12th Pass,"Rs 1,00,14,348 ~ 1 Crore+",Rs 0 ~
3,4,Abdur Rahim Boxi,MALATIPUR,AITC,2,12th Pass,"Rs 5,82,83,056 ~ 5 Crore+","Rs 10,49,505 ~ 10 Lacs+"
4,5,Adhikari Dibyendu,EGRA,BJP,2,Graduate,"Rs 4,80,05,224 ~ 4 Crore+","Rs 50,59,168 ~ 50 Lacs+"
...,...,...,...,...,...,...,...,...
287,288,Tusar Kumar Majumdar,PANDUA,BJP,2,12th Pass,"Rs 3,15,68,724 (~3 Crore+)",Rs 0
288,289,Umesh Rai,HOWRAH UTTAR,BJP,19,Graduate,"Rs 76,17,624 ~ 76 Lacs+",Rs 0 ~
289,290,Usha Rani Mondal,MINAKHAN (SC),AITC,0,8th Pass,"Rs 2,37,58,335 ~ 2 Crore+","Rs 19,68,000 ~ 19 Lacs+"
290,291,Utpal Brahmacharo,KALIAGANJ (SC),BJP,0,Graduate,"Rs 44,34,501 ~ 44 Lacs+",Rs 0 ~


In [ ]:
# @title
df['Assets_Amount'] = (
    df['Assets']
      .str.split(r'~|\(', regex=True)
      .str[0]
      .str.replace('Rs', '', regex=False)
      .str.replace(',', '', regex=False)
      .str.strip()
      .astype(int)
)

df['Liabilities_Amount'] = (
    df['Liabilities']
      .str.split(r'~|\(', regex=True)
      .str[0]
      .str.replace('Rs', '', regex=False)
      .str.replace(',', '', regex=False)
      .str.strip()
      .astype(int)
)

df.drop(columns=['Assets', 'Liabilities'], inplace=True)
df

,Sno,Candidate,Constituency,Party,Criminal Cases,Education,Assets_Amount,Liabilities_Amount
0,1,Abdul Aziz Doctor,LALGOLA,AITC,0,Post Graduate,62908242,3850000
1,2,Abdul Khaleque Molla,METIABURUZ,AITC,2,10th Pass,17455505,0
2,3,Abdul Matin Muhammad,HAROA,AITC,1,12th Pass,10014348,0
3,4,Abdur Rahim Boxi,MALATIPUR,AITC,2,12th Pass,58283056,1049505
4,5,Adhikari Dibyendu,EGRA,BJP,2,Graduate,48005224,5059168
...,...,...,...,...,...,...,...,...
287,288,Tusar Kumar Majumdar,PANDUA,BJP,2,12th Pass,31568724,0
288,289,Umesh Rai,HOWRAH UTTAR,BJP,19,Graduate,7617624,0
289,290,Usha Rani Mondal,MINAKHAN (SC),AITC,0,8th Pass,23758335,1968000
290,291,Utpal Brahmacharo,KALIAGANJ (SC),BJP,0,Graduate,4434501,0


In [ ]:
count=df['Candidate'].value_counts()
candidates=count[count==2].index

constituency_2=df[df['Candidate'].isin(candidates)]

In [ ]:
df.loc[df['Sno'] == 32, ['Criminal Cases', 'Education']] = [3, 'Post Graduate',]
df.loc[df['Sno'] == 31, ['Assets_Amount', 'Liabilities_Amount']] = [39036176, 8409674,]

In [ ]:
df['Constituency'] = (
    df.groupby('Candidate')['Constituency']
    .transform(lambda x: ', '.join(x))
)

df = df.drop_duplicates(subset='Candidate')


In [ ]:
df["Education"] = df["Education"].replace(
    "Graduate Professional",
    "Graduate"
)

df["Education"].value_counts()

/tmp/ipykernel_610/3088768229.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



,count
Education,
Graduate,111
Post Graduate,65
12th Pass,46
10th Pass,27
8th Pass,17
Others,8
Doctorate,7
Literate,5
5th Pass,2


In [ ]:
lower_education = [
    '12th Pass',
    '10th Pass',
    '8th Pass',
    'Literate',
    '5th Pass',
    'Illiterate'
]

higher_education = [
    'Graduate',
    'Post Graduate',
    'Doctorate'
]

df["Education_level"] = "Others"

df.loc[
    df["Education"].isin(lower_education),
    "Education_level"
] = "Lower Education"

df.loc[
    df["Education"].isin(higher_education),
    "Education_level"
] = "Higher Education"

df["Education_level"].value_counts()

/tmp/ipykernel_610/3803533011.py:16: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



,count
Education_level,
Higher Education,183
Lower Education,98
Others,8


In [ ]:
#cleaned data
df

,Sno,Candidate,Constituency,Party,Criminal Cases,Education,Assets_Amount,Liabilities_Amount,Education_level
0,1,Abdul Aziz Doctor,LALGOLA,AITC,0,Post Graduate,62908242,3850000,Higher Education
1,2,Abdul Khaleque Molla,METIABURUZ,AITC,2,10th Pass,17455505,0,Lower Education
2,3,Abdul Matin Muhammad,HAROA,AITC,1,12th Pass,10014348,0,Lower Education
3,4,Abdur Rahim Boxi,MALATIPUR,AITC,2,12th Pass,58283056,1049505,Lower Education
4,5,Adhikari Dibyendu,EGRA,BJP,2,Graduate,48005224,5059168,Higher Education
...,...,...,...,...,...,...,...,...,...
287,288,Tusar Kumar Majumdar,PANDUA,BJP,2,12th Pass,31568724,0,Lower Education
288,289,Umesh Rai,HOWRAH UTTAR,BJP,19,Graduate,7617624,0,Higher Education
289,290,Usha Rani Mondal,MINAKHAN (SC),AITC,0,8th Pass,23758335,1968000,Lower Education
290,291,Utpal Brahmacharo,KALIAGANJ (SC),BJP,0,Graduate,4434501,0,Higher Education


In [ ]:
df['net_worth']=df['Assets_Amount']-df['Liabilities_Amount']
df.sort_values('net_worth', ascending=False).head(20)

/tmp/ipykernel_610/1510671323.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



,Sno,Candidate,Constituency,Party,Criminal Cases,Education,Assets_Amount,Liabilities_Amount,Education_level,net_worth
84,85,Dilip Saha,NABAGRAM (SC),BJP,0,8th Pass,430396618,0,Lower Education,430396618
206,207,Ritesh Tiwari,KASHIPUR-BELGACHHIA,BJP,4,Graduate,278898708,0,Higher Education,278898708
276,277,Swaraj Ghosh,SAPTAGRAM,BJP,4,10th Pass,289863958,23879791,Lower Education,265984167
167,168,Nandita Chowdhury,HOWRAH DAKSHIN,AITC,0,12th Pass,264416942,7009495,Lower Education,257407447
46,47,Bayron Biswas,SAGARDIGHI,AITC,1,12th Pass,334222210,79649669,Lower Education,254572541
154,155,Md. Matibur Rahaman,HARISCHANDRAPUR,AITC,2,8th Pass,256819938,27282547,Lower Education,229537391
103,104,Firhad Hakim,KOLKATA PORT,AITC,2,Graduate,225753159,13679175,Higher Education,212073984
160,161,Mohammed Noor Alam,SAMSERGANJ,AITC,0,12th Pass,222321497,16833659,Lower Education,205487838
180,181,Partha Ghosh,RANIGANJ,BJP,0,Post Graduate,177664006,3513888,Higher Education,174150118
235,236,Sharadwat Mukherjee,BIDHANNAGAR,BJP,0,Post Graduate,157226453,4830854,Higher Education,152395599


In [ ]:
# @title 1. Top 20 Candidates by Declared Net Worth (Interactive Dashboard)

import pandas as pd
import plotly.graph_objects as go

# ==========================================================
# DATA PREPARATION
# ==========================================================

df["net_worth"] = df["Assets_Amount"] - df["Liabilities_Amount"]

top20 = (
    df.sort_values("net_worth", ascending=False)
      .head(20)
      .copy()
)

top20["NetWorthCr"] = (top20["net_worth"] / 1e7).round(2)

# ==========================================================
# PARTY COLOURS
# ==========================================================

party_colors = {
    "BJP": "#FF9933",
    "AITC": "#009E60",
    "INC": "#1976D2",
    "CPI(M)": "#D32F2F",
    "All India Secular Front": "#7B1FA2",
    "Aam Janata Unnayan party": "#795548"
}

top20["Color"] = (
    top20["Party"]
    .map(party_colors)
    .fillna("#90A4AE")
)

# ==========================================================
# FIGURE
# ==========================================================

fig = go.Figure()

# ----------------------------------------------------------
# Main Bar Chart
# ----------------------------------------------------------

fig.add_trace(

    go.Bar(

        x=top20["Candidate"],

        y=top20["NetWorthCr"],

        marker=dict(

            color=top20["Color"],

            line=dict(
                color="black",
                width=1
            )
        ),

        text=[f"₹{v:.1f} Cr" for v in top20["NetWorthCr"]],

        textposition="outside",

        customdata=top20[
            [
                "Party",
                "Constituency",
                "Education",
                "Assets_Amount",
                "Liabilities_Amount"
            ]
        ],

        hovertemplate=
        "<b>%{x}</b><br><br>"
        "<b>Party:</b> %{customdata[0]}<br>"
        "<b>Constituency:</b> %{customdata[1]}<br>"
        "<b>Education:</b> %{customdata[2]}<br><br>"
        "<b>Declared Assets:</b> ₹%{customdata[3]:,.0f}<br>"
        "<b>Declared Liabilities:</b> ₹%{customdata[4]:,.0f}<br>"
        "<b>Net Worth:</b> ₹%{y:.2f} Cr"
        "<extra></extra>",

        showlegend=False
    )
)

# ==========================================================
# PARTY LEGEND
# ==========================================================

for party, color in party_colors.items():

    fig.add_trace(

        go.Scatter(

            x=[None],
            y=[None],

            mode="markers",

            marker=dict(
                size=12,
                color=color
            ),

            name=party
        )
    )

# ==========================================================
# TOP 5 PANEL
# ==========================================================

top5 = top20.head(5)

summary = "<b>🏆 TOP 5 CANDIDATES</b><br><br>"

for i, (_, row) in enumerate(top5.iterrows(), 1):

    summary += (
        f"<span style='color:{party_colors.get(row['Party'],'black')}'>●</span> "
        f"<b>#{i} {row['Candidate']}</b><br>"
        f"&nbsp;&nbsp;{row['Party']}<br>"
        f"&nbsp;&nbsp;{row['Constituency']}<br>"
        f"&nbsp;&nbsp;{row['Education']}<br><br>"
    )

# ==========================================================
# LAYOUT
# ==========================================================

fig.update_layout(

    template="plotly_white",

    width=1500,

    height=850,

    title=dict(

        text=(
            "<b>WEST BENGAL ASSEMBLY ELECTION 2026</b><br>"
            "<span style='font-size:20px;'>"
            "Top 20 Candidates by Declared Net Worth"
            "</span>"
        ),

        x=0.5,

        font=dict(
            family="Arial",
            size=28,
            color="#0D1B2A"
        )
    ),

    xaxis=dict(

        title="",

        tickangle=-65,

        tickfont=dict(size=10)

    ),

    yaxis=dict(

        title="<b>Declared Net Worth (₹ Crore)</b>",

        gridcolor="rgba(0,0,0,0.08)",

        zeroline=False

    ),

    paper_bgcolor="white",

    plot_bgcolor="white",

    legend=dict(

        title="<b>Political Party</b>",

        orientation="h",

        y=.95,

        x=0.5,

        xanchor="center",

        font=dict(size=11)

    ),

    margin=dict(
        l=70,
        r=320,
        t=120,
        b=170
    )

)

# ==========================================================
# TOP 5 BOX
# ==========================================================

fig.add_annotation(

    x=1.02,
    y=1.05,

    xref="paper",
    yref="paper",

    text=summary,

    showarrow=False,

    align="left",

    bordercolor="#B0BEC5",

    borderwidth=1,

    borderpad=8,

    bgcolor="#F8F9FA",

    font=dict(size=11)

)

# ==========================================================
# SOURCE
# ==========================================================

fig.add_annotation(

    x=0,
    y=-0.18,

    xref="paper",
    yref="paper",

    text=(
        "<i>Source: MyNeta (ADR & National Election Watch) • "
        "Net Worth = Declared Assets − Declared Liabilities</i>"
    ),

    showarrow=False,

    font=dict(
        size=11,
        color="gray"
    )

)

fig.show()

/tmp/ipykernel_610/2614676619.py:10: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



### 💰 Key Insight — Top 20 Candidates by Declared Net Worth

The wealthiest winners are concentrated among **BJP and AITC**, with BJP's **Dilip Saha** standing out at approximately **₹43.0 Cr**, substantially ahead of the next-ranked candidate, **Ritesh Tiwari (₹27.9 Cr)**. Among the top 20, **BJP and AITC dominate the high-net-worth segment**, while INC appears only once. This indicates a notable concentration of substantial declared wealth among candidates from the two leading parties, although **net worth reflects financial position and should not be interpreted as electoral strength or political influence by itself**.

In [ ]:

# @title 2. What is the education distribution among candidates?
import pandas as pd
import plotly.graph_objects as go
import plotly.express as px

# ==========================================================
# EDUCATION DISTRIBUTION
# ==========================================================

education_distribution = (
    df["Education"]
    .value_counts()
    .reset_index()
)

education_distribution.columns = ["Education", "Count"]

education_distribution["Percentage"] = (
    education_distribution["Count"]
    / education_distribution["Count"].sum()
    * 100
).round(1)

total_candidates = education_distribution["Count"].sum()

# ==========================================================
# BLUE GRADIENT (DARKER = MORE CANDIDATES)
# ==========================================================

colors = px.colors.sequential.Blues[::-1]

if len(education_distribution) <= len(colors):
    pie_colors = colors[:len(education_distribution)]
else:
    pie_colors = px.colors.sample_colorscale(
        "Blues",
        [i/(len(education_distribution)-1)
         for i in range(len(education_distribution))]
    )

# ==========================================================
# LABELS
# Show percentage only if >= 5%
# ==========================================================

labels = [
    f"{p:.1f}%"
    if p >= 5
    else ""
    for p in education_distribution["Percentage"]
]

# ==========================================================
# DONUT CHART
# ==========================================================

fig = go.Figure()

fig.add_trace(

    go.Pie(

        labels=education_distribution["Education"],

        values=education_distribution["Count"],

        hole=0.60,

        marker=dict(
            colors=pie_colors,
            line=dict(color="white", width=2)
        ),

        text=labels,

        textinfo="text",

        textfont=dict(
            size=15,
            color="black"
        ),

        hovertemplate=
        "<b>%{label}</b><br><br>"
        "Candidates : <b>%{value}</b><br>"
        "Percentage : <b>%{percent}</b>"
        "<extra></extra>",

        sort=False
    )
)

# ==========================================================
# CENTER TEXT
# ==========================================================

fig.add_annotation(

    x=0.5,
    y=0.5,

    text=
    f"<b>{total_candidates}</b><br>"
    "<span style='font-size:18px'>Candidates</span>",

    showarrow=False,

    font=dict(
        size=30,
        color="#0D47A1"
    )
)

# ==========================================================
# TITLE
# ==========================================================

fig.update_layout(

   title=dict(
    text=(
        "<b>WEST BENGAL ASSEMBLY ELECTION 2026</b><br>"
        "<span style='font-size:20px;'>Educational Profile of Elected Members of the Legislative Assembly (MLAs)</span>"
    ),
    x=0.5,
    xanchor="center",
    font=dict(
        family="Arial",
        size=28,
        color="#0D1B2A"
    )
),

    legend=dict(

        title="<b>Education Level</b>",

        font=dict(size=13),

        x=1.02,

        y=0.5,

        bgcolor="rgba(255,255,255,0.9)"
    ),

    width=1000,

    height=700,

    paper_bgcolor="white",

    plot_bgcolor="white",

    margin=dict(
        l=20,
        r=150,
        t=90,
        b=30
    )
)

fig.show()

### 🎓 Key Insight — Educational Profile of Elected MLAs

The educational profile is strongly concentrated at the **higher-education end**: **38.4% of elected MLAs are Graduates**, while **22.5% are Post Graduates** and **15.9% are 12th Pass**. Together, **more than 60% of the elected representatives are Graduates or Post Graduates**, indicating a predominantly higher-educated legislative profile. Lower educational qualifications form relatively smaller shares, suggesting that **formal educational attainment among elected representatives is generally high**.

In [ ]:
# @title 3. Average assets and liabilities of competitive parties.
import plotly.graph_objects as go
import pandas as pd

# ==========================================================
# DATA PREPARATION
# ==========================================================

# Keep only parties with >=5 seats
party_counts = df["Party"].value_counts()
major_parties = party_counts[party_counts >= 5].index

filtered = df[df["Party"].isin(major_parties)]

summary = (
    filtered.groupby("Party")[["Assets_Amount","Liabilities_Amount"]]
    .mean()/1e7
).round(2)

summary = summary.reset_index()

# ==========================================================
# PARTY COLORS
# ==========================================================

party_colors = {
    "BJP":"#FF9933",
    "AITC":"#00A651",
    "INC":"#1976D2",
    "CPI(M)":"#D32F2F",
    "All India Secular Front":"#7B1FA2",
    "Aam Janata Unnayan party":"#795548"
}

axis_colors = [
    party_colors.get(p,"black")
    for p in summary["Party"]
]

# ==========================================================
# FIGURE
# ==========================================================

fig = go.Figure()

# Assets
fig.add_trace(
    go.Bar(
        x=summary["Party"],
        y=summary["Assets_Amount"],
        name="Average Assets",
        marker_color="#1565C0",
        text=[f"₹{x:.1f} Cr" for x in summary["Assets_Amount"]],
        textposition="outside",
        hovertemplate=
        "<b>%{x}</b><br>"
        "Average Assets: ₹%{y:.2f} Cr"
        "<extra></extra>"
    )
)

# Liabilities
fig.add_trace(
    go.Bar(
        x=summary["Party"],
        y=summary["Liabilities_Amount"],
        name="Average Liabilities",
        marker_color="#D32F2F",
        text=[f"₹{x:.1f} Cr" for x in summary["Liabilities_Amount"]],
        textposition="outside",
        hovertemplate=
        "<b>%{x}</b><br>"
        "Average Liabilities: ₹%{y:.2f} Cr"
        "<extra></extra>"
    )
)

# ==========================================================
# SEAT DISTRIBUTION BOX
# ==========================================================

seat_text = "<b>Party Seat Distribution</b><br><br>"

for party,seats in party_counts.items():

    color = party_colors.get(party,"black")

    seat_text += (
        f"<span style='color:{color}'>●</span> "
        f"<b>{party}</b> : {seats}<br>"
    )

# ==========================================================
# LAYOUT
# ==========================================================

fig.update_layout(

    template="plotly_white",

    barmode="group",

    width=1200,

    height=700,

    title=dict(
        text=(
            "<b>WEST BENGAL ASSEMBLY ELECTION 2026</b><br>"
            "<sup>Party-wise Average Declared Assets & Liabilities of Elected MLAs</sup>"
        ),
        x=0.5,
        font=dict(size=24)
    ),

    yaxis_title="<b>Average Value (₹ Crore)</b>",

    xaxis_title="",

    font=dict(
        family="Arial",
        size=14
    ),

    legend=dict(
        orientation="h",
        y=1.08,
        x=0.5,
        xanchor="center"
    ),

    margin=dict(
        l=60,
        r=250,
        t=110,
        b=70
    )
)

# ==========================================================
# PARTY COLORED X LABELS
# ==========================================================

fig.update_xaxes(
    tickfont=dict(size=13)
)

# ==========================================================
# SIDE PANEL
# ==========================================================

fig.add_annotation(

    x=1.18,
    y=0.98,

    xref="paper",
    yref="paper",

    align="left",

    showarrow=False,

    bordercolor="black",

    borderwidth=1,

    borderpad=8,

    bgcolor="#F8F9FA",

    text=seat_text
)

# ==========================================================
# FOOTNOTE
# ==========================================================

fig.add_annotation(

    x=0,
    y=-0.18,

    xref="paper",
    yref="paper",

    showarrow=False,

    align="left",

    text=(
        "<i>Source: MyNeta (ADR & National Election Watch). "
        "Values represent average declared assets and liabilities of elected candidates.</i>"
    )
)

fig.show()

### 💰 Key Insight — Party-wise Financial Profile of Elected MLAs

AITC elected MLAs report a higher average asset base of **₹5.3 Cr**, compared with **₹3.2 Cr** for BJP MLAs. AITC also records higher average liabilities (**₹1.1 Cr**) than BJP (**₹0.5 Cr**). Despite BJP having a much larger legislative presence (**204 seats vs 80**), its average declared assets per elected MLA are lower than those of AITC.

Overall, the chart highlights a clear difference in the **financial profiles of elected representatives across the two major parties**, with AITC showing higher average assets as well as higher average liabilities.

In [ ]:
# @title 4. Maximum Declared Assets vs Maximum Declared Liabilities by Political Party

import plotly.graph_objects as go

# ==========================================================
# DATA
# ==========================================================

summary = (
    df.groupby("Party")[["Assets_Amount", "Liabilities_Amount"]]
      .max()
      .reset_index()
)

summary["AssetsCr"] = (summary["Assets_Amount"] / 1e7).round(2)
summary["LiabilitiesCr"] = (summary["Liabilities_Amount"] / 1e7).round(2)

summary = summary.sort_values("AssetsCr", ascending=True)

# ==========================================================
# FIGURE
# ==========================================================

fig = go.Figure()

# ----------------------------------------------------------
# CONNECTING LINES
# ----------------------------------------------------------

for _, row in summary.iterrows():

    fig.add_trace(
        go.Scatter(
            x=[row["LiabilitiesCr"], row["AssetsCr"]],
            y=[row["Party"], row["Party"]],
            mode="lines",
            line=dict(color="#BDBDBD", width=4),
            hoverinfo="skip",
            showlegend=False
        )
    )

# ----------------------------------------------------------
# ASSET MARKERS
# ----------------------------------------------------------

fig.add_trace(

    go.Scatter(

        x=summary["AssetsCr"],

        y=summary["Party"],

        mode="markers",

        name="Maximum Assets",

        marker=dict(
            size=18,
            color="#1565C0",
            line=dict(color="white", width=2)
        ),

        hovertemplate=
        "<b>%{y}</b><br>"
        "Maximum Declared Assets<br>"
        "₹%{x:.2f} Cr"
        "<extra></extra>"
    )
)

# ----------------------------------------------------------
# LIABILITY MARKERS
# ----------------------------------------------------------

fig.add_trace(

    go.Scatter(

        x=summary["LiabilitiesCr"],

        y=summary["Party"],

        mode="markers",

        name="Maximum Liabilities",

        marker=dict(
            size=18,
            color="#D32F2F",
            line=dict(color="white", width=2)
        ),

        hovertemplate=
        "<b>%{y}</b><br>"
        "Maximum Declared Liabilities<br>"
        "₹%{x:.2f} Cr"
        "<extra></extra>"
    )
)


# ==========================================================
# SMART LABELS (CORRECTED FOR ANY ASSET/LIABILITY ORDER)
# ==========================================================

for _, row in summary.iterrows():
    asset = row["AssetsCr"]
    liability = row["LiabilitiesCr"]
    gap = abs(asset - liability)

    # 1. VERY CLOSE VALUES -> Stack vertically
    if gap < 0.5:
        # Higher value goes slightly up, lower value goes slightly down
        asset_yshift = 16 if asset >= liability else -16
        liab_yshift = -16 if asset >= liability else 16

        fig.add_annotation(
            x=asset,
            y=row["Party"],
            text=f"<b>  ₹{asset:.1f} Cr  </b>",
            showarrow=False,
            yshift=asset_yshift,
            font=dict(color="#1565C0", size=12)
        )

        fig.add_annotation(
            x=liability,
            y=row["Party"],
            text=f"<b>  ₹{liability:.1f} Cr  </b>",
            showarrow=False,
            yshift=liab_yshift,
            font=dict(color="#D32F2F", size=12)
        )

    # 2. NORMAL / FAR APART VALUES -> Dynamic Left/Right shift
    else:
        # Asset label position
        asset_xshift = 50 if asset > liability else -50
        fig.add_annotation(
            x=asset,
            y=row["Party"],
            text=f"<b>  ₹{asset:.1f} Cr  </b>",
            showarrow=False,
            xshift=asset_xshift,
            font=dict(color="#1565C0", size=12)
        )

        # Liability label position
        liab_xshift = -50 if asset > liability else 50
        fig.add_annotation(
            x=liability,
            y=row["Party"],
            text=f"<b>  ₹{liability:.1f} Cr  </b>",
            showarrow=False,
            xshift=liab_xshift,
            font=dict(color="#D32F2F", size=12)
        )


# ==========================================================
# LAYOUT ADJUSTMENTS (Fixes Legend Overlap)
# ==========================================================

fig.update_layout(
    template="plotly_white",
    width=1200,
    height=700,
    title=dict(
        text=(
            "<b>WEST BENGAL ASSEMBLY ELECTION 2026</b><br>"
            "<span style='font-size:18px; color:#555555'>"
            "Maximum Declared Assets vs Maximum Declared Liabilities by Political Party"
            "</span>"
        ),
        x=0.5,
        y=0.96, # Moved title slightly down
        font=dict(size=24, color="#0D1B2A")
    ),
    xaxis=dict(
        title="<b>Amount (₹ Crore)</b>",
        showgrid=True,
        gridcolor="rgba(0,0,0,0.08)",
        zeroline=False
    ),
    yaxis=dict(
        title="",
        categoryorder="array",
        categoryarray=summary["Party"]
    ),
    legend=dict(
        orientation="h",
        x=0.5,
        xanchor="center",
        y=1.02 # Adjusted legend so it doesn't collide with title
    ),
    paper_bgcolor="white",
    plot_bgcolor="white",
    margin=dict(
        l=150, # Slightly extra room for long party names
        r=80,
        t=140, # Extra top margin for title & legend
        b=60
    )
)

fig.show()

In [ ]:
# @title
richest = (
    df.loc[df.groupby("Party")["Assets_Amount"].idxmax()]
    [["Party", "Candidate", "Assets_Amount"]]
    .rename(columns={
        "Candidate": "Highest Asset Candidate",
        "Assets_Amount": "Max Assets"
    })
)

liability = (
    df.loc[df.groupby("Party")["Liabilities_Amount"].idxmax()]
    [["Party", "Candidate", "Liabilities_Amount"]]
    .rename(columns={
        "Candidate": "Highest Liability Candidate",
        "Liabilities_Amount": "Max Liabilities"
    })
)

summary = richest.merge(liability, on="Party")

summary

,Party,Highest Asset Candidate,Max Assets,Highest Liability Candidate,Max Liabilities
0,AITC,Ahmed Javed Khan,395353826,Ahmed Javed Khan,437240034
1,Aam Janata Unnayan party,Humayun Kabir,44414849,Humayun Kabir,0
2,All India Secular Front,Md Nawsad Siddique,5569232,Md Nawsad Siddique,0
3,BJP,Dilip Saha,430396618,Dipankar Jana,102744080
4,CPI(M),Md. Mostafijur Rahaman,35214800,Md. Mostafijur Rahaman,0
5,INC,Julfikar Ali,188292606,Motab Shaikh,135927098


### 💰 Key Insight — Maximum Assets vs Maximum Liabilities by Party

The financial profile varies considerably across parties. **BJP records the highest declared asset value at ₹43.0 Cr**, while its maximum declared liabilities stand at approximately **₹10.3 Cr**. AITC follows with a maximum asset value of **₹39.5 Cr**, but records the **highest maximum liabilities at ₹43.7 Cr**. INC shows a comparatively smaller maximum asset value of **₹18.8 Cr** and maximum liabilities of **₹13.6 Cr**.

Among the smaller parties, maximum declared assets remain relatively low: **Aam Janata Unnayan Party (₹4.4 Cr), CPI(M) (₹3.5 Cr), and All India Secular Front (₹0.6 Cr)**. Their reported maximum liabilities are either negligible or zero in this dataset.

### ⚠️ Notable Financial Flags

The most striking case is **AITC**, where maximum declared liabilities (**₹43.7 Cr**) exceed maximum declared assets (**₹39.5 Cr**) by approximately **₹4.2 Cr**. This produces an unusual liability-to-asset position and warrants verification of the underlying affidavit and the **composition and nature of those liabilities**.

However, this should **not automatically be interpreted as financial loss, tax avoidance, or wrongdoing**. The chart only identifies an unusual financial position; further investigation of the original affidavit would be required to understand it.

### 📌 Analytical Takeaway

The chart reveals a clear contrast between **asset concentration and liability exposure across parties**: BJP has the highest single declared asset value, while AITC has the highest single declared liability value. The AITC case stands out as the primary **financial due-diligence flag**, whereas the remaining parties show substantially smaller maximum financial exposures.

In [ ]:
# @title 5. Educational Distribution of Elected MLAs by Political Party
# ==========================================================
# Professional Party-wise Education Donut Dashboard
# ==========================================================

import math
import pandas as pd
from plotly.subplots import make_subplots
import plotly.graph_objects as go

# ----------------------------------------------------------
# Prepare Data
# ----------------------------------------------------------

party_education = (
    df.groupby(["Party", "Education"])
      .size()
      .reset_index(name="Count")
)

# Parties
parties = sorted(party_education["Party"].unique())

# Total candidates in each party
party_total = (
    party_education.groupby("Party")["Count"]
    .sum()
    .to_dict()
)

# ----------------------------------------------------------
# Education Order (optional)
# ----------------------------------------------------------

education_order = [
    "Illiterate",
    "Literate",
    "5th Pass",
    "8th Pass",
    "10th Pass",
    "12th Pass",
    "Graduate",
    "Graduate Professional",
    "Post Graduate",
    "Doctorate",
    "Others"
]

# Keep only available education levels
education_order = [
    e for e in education_order
    if e in party_education["Education"].unique()
]

# ----------------------------------------------------------
# Blue Shades
# ----------------------------------------------------------

education_colors = {
    "Illiterate": "#F7FBFF",
    "Literate": "#E3F2FD",
    "5th Pass": "#BBDEFB",
    "8th Pass": "#90CAF9",
    "10th Pass": "#64B5F6",
    "12th Pass": "#42A5F5",
    "Graduate": "#1E88E5",
    "Graduate Professional": "#1976D2",
    "Post Graduate": "#1565C0",
    "Doctorate": "#0D47A1",
    "Others": "#9E9E9E"
}

# ----------------------------------------------------------
# Layout
# ----------------------------------------------------------

cols = 3
rows = math.ceil(len(parties) / cols)

subplot_titles = [f"<b>{p}</b>" for p in parties]

fig = make_subplots(
    rows=rows,
    cols=cols,
    specs=[[{'type':'domain'}]*cols for _ in range(rows)],
    vertical_spacing=0.12,
    horizontal_spacing=0.08
)

# ----------------------------------------------------------
# Add Donuts
# ----------------------------------------------------------

for i, party in enumerate(parties):

    r = i // cols + 1
    c = i % cols + 1

    temp = (
    party_education[party_education["Party"] == party]
    .set_index("Education")
    .reindex(education_order, fill_value=0)
    .reset_index()
    .query("Count > 0")
)

    fig.add_trace(

        go.Pie(

            labels=temp["Education"],
            values=temp["Count"],

            hole=0.65,

            sort=False,

            marker=dict(
                colors=[education_colors.get(x, "#CCCCCC") for x in temp["Education"]],
                line=dict(color="white", width=2)
            ),

            textinfo="percent",

            textfont_size=11,

            hovertemplate=
            "<b>%{label}</b><br>" +
            "Candidates: %{value}<br>" +
            "Percentage: %{percent}<extra></extra>",

            showlegend=(i==0)

        ),

        row=r,
        col=c

    )

# ----------------------------------------------------------
# Candidate Count in Centre
# ----------------------------------------------------------

for i, trace in enumerate(fig.data):

    party = parties[i]

    x = sum(trace.domain["x"]) / 2
    y = sum(trace.domain["y"]) / 2

    fig.add_annotation(

        x=x,
        y=y,

        xref="paper",
        yref="paper",

        text=f"<b>{party_total[party]}</b>",

        showarrow=False,

        font=dict(
            size=24,
            color="black"
        )

    )
# ----------------------------------------------------------
# Party Name Below Each Donut
# ----------------------------------------------------------

for i, trace in enumerate(fig.data):

    party = parties[i]

    x = sum(trace.domain["x"]) / 2
    y = trace.domain["y"][0] - 0.035   # Distance below donut

    fig.add_annotation(
        x=x,
        y=y,
        xref="paper",
        yref="paper",
        text=f"<b>{party}</b>",
        showarrow=False,
        font=dict(
            size=16,
            color="black"
        ),
        align="center"
    )
# ----------------------------------------------------------
# Layout
# ----------------------------------------------------------

fig.update_layout(

    title=dict(
        text="<b>Educational Representation Across Political Parties</b>",
        x=0.5,
        font=dict(size=28)
    ),

    height=420*rows,
    width=1300,

    paper_bgcolor="white",
    plot_bgcolor="white",

    font=dict(
        family="Arial",
        size=13
    ),

    legend_title="Education",

    margin=dict(
        t=90,
        l=40,
        r=40,
        b=80
    )
)

# ----------------------------------------------------------
# Make subplot titles larger
# ----------------------------------------------------------

for ann in fig.layout.annotations:
    ann.font.size = 16

# ----------------------------------------------------------
# Show
# ----------------------------------------------------------

fig.show()

### 🎓 Key Insight — Educational Representation Across Political Parties

The educational profiles of the two major parties, **BJP and AITC**, are broadly similar at the top level: **Graduates form the largest group in both parties**, accounting for **39.7% of BJP MLAs and 38.3% of AITC MLAs**. However, their profiles differ beyond graduation. **BJP has a larger share of Post Graduates (10.8%)**, while AITC has a considerably larger proportion of **10th- and 12th-pass MLAs**, each accounting for **23.8%** of its elected members.

BJP also shows a broader representation across educational categories, including **Doctorate, Post Graduate, Graduate, 12th Pass, 10th Pass and lower qualifications**, whereas AITC's representation is more concentrated around **Graduate, 12th Pass and 10th Pass** qualifications.

### 📌 Important Context

The smaller parties should **not be directly compared with BJP and AITC on percentages**, because their sample sizes are extremely small. Aam Janata Unnayan Party, All India Secular Front and CPI(M) each have **only 1 elected member**, while INC has **2**. A 100% share in these groups therefore represents just one candidate rather than a broader party-level educational pattern.

**Overall, the major-party comparison suggests broadly similar graduate representation, but BJP has a somewhat stronger presence of higher postgraduate qualifications, while AITC has greater representation among 10th- and 12th-pass candidates.**

In [ ]:
# @title
zero_cases = (
    df[df["Criminal Cases"] == 0]
    .groupby("Party")
    .size()
    .rename("Zero Criminal Count")
    .to_frame()
)

zero_cases["Total Winners"] = df.groupby("Party").size()

zero_cases["Zero Criminal (%)"] = (
    zero_cases["Zero Criminal Count"]
    / zero_cases["Total Winners"]
    * 100
).round(2)

zero_cases

,Zero Criminal Count,Total Winners,Zero Criminal (%)
Party,,,
AITC,46,80,57.50
BJP,53,204,25.98
INC,2,2,100.00


In [ ]:
# @title 6. How Clean Are the Winning Candidates?
import pandas as pd
import plotly.graph_objects as go

# ==========================================================
# Prepare Data
# ==========================================================

party_summary = (
    df.groupby("Party")
      .agg(
          Total_Winners=("Party", "size"),
          Zero_Criminal=("Criminal Cases", lambda x: (x == 0).sum())
      )
)

party_summary["One_or_More_Criminal"] = (
    party_summary["Total_Winners"] - party_summary["Zero_Criminal"]
)

party_summary["Zero Criminal (%)"] = (
    party_summary["Zero_Criminal"] /
    party_summary["Total_Winners"] * 100
).round(1)

party_summary["1+ Criminal (%)"] = (
    party_summary["One_or_More_Criminal"] /
    party_summary["Total_Winners"] * 100
).round(1)

party_summary = (
    party_summary
    .sort_values("Zero Criminal (%)", ascending=True)
    .reset_index()
)

# ==========================================================
# Party Colors
# (Modify if your dashboard uses different colors)
# ==========================================================

party_colors = {
    "AITC": "#00A651",      # Green
    "BJP": "#F57C00",       # Saffron
    "INC": "#19A0D8",       # Blue
    "CPI(M)": "#D32F2F",    # Red
    "RSP": "#8B0000",
    "AIFB": "#C62828",
    "ISF": "#00897B",
    "IND": "#757575"
}

default_color = "#607D8B"

# ==========================================================
# Create Figure
# ==========================================================

fig = go.Figure()

for _, row in party_summary.iterrows():

    party = row["Party"]

    color = party_colors.get(
        party,
        default_color
    )

    zero_pct = row["Zero Criminal (%)"]
    criminal_pct = row["1+ Criminal (%)"]

    # ----------------------------
    # Zero Criminal Cases
    # ----------------------------

    fig.add_trace(

        go.Bar(

            y=[party],

            x=[zero_pct],

            orientation="h",

            marker=dict(
                color=color
            ),

            name="Zero Criminal Cases",

            text=[
                f"{zero_pct:.1f}%"
                if zero_pct >= 8 else ""
            ],

            textposition="inside",

            textfont=dict(
                color="white",
                size=13
            ),

            customdata=[
                [
                    row["Zero_Criminal"],
                    row["Total_Winners"]
                ]
            ],

            hovertemplate=
            "<b>%{y}</b><br>"
            "Zero Criminal Winners: %{customdata[0]}<br>"
            "Total Winners: %{customdata[1]}<br>"
            "Percentage: %{x}%<extra></extra>",

            showlegend=False

        )

    )

    # ----------------------------
    # Criminal Cases
    # ----------------------------

    fig.add_trace(

        go.Bar(

            y=[party],

            x=[criminal_pct],

            orientation="h",

            marker=dict(
                color="black"
            ),

            name="1 or More Criminal Cases",

            text=[
                f"{criminal_pct:.1f}%"
                if criminal_pct >= 8 else ""
            ],

            textposition="inside",

            textfont=dict(
                color="white",
                size=13
            ),

            hovertemplate=
            "<b>%{y}</b><br>"
            "1 or More Criminal Cases: %{x}%<extra></extra>",

            showlegend=False

        )

    )

# ==========================================================
# Layout
# ==========================================================

fig.update_layout(

    barmode="stack",

    title=dict(

        text="<b>Winning with a Clean Slate?</b>",

        x=0.5,

        font=dict(
            size=26
        )

    ),

    xaxis=dict(

        title="Share of Winning Candidates",

        range=[0, 100],

        ticksuffix="%",

        showgrid=True,

        gridcolor="#E5E5E5"

    ),

    yaxis=dict(

        title="Political Party",

        categoryorder="array",

        categoryarray=party_summary["Party"]

    ),

    height=700,

    width=1150,

    paper_bgcolor="white",

    plot_bgcolor="white",

    font=dict(

        family="Arial",

        size=13

    ),

    margin=dict(

        l=90,

        r=50,

        t=90,

        b=60

    )

)

# ==========================================================
# Manual Legend
# ==========================================================

fig.add_trace(
    go.Bar(
        x=[None],
        y=[None],
        marker=dict(color="#4CAF50"),
        name="Zero Criminal Cases"
    )
)

fig.add_trace(
    go.Bar(
        x=[None],
        y=[None],
        marker=dict(color="black"),
        name="1 or More Criminal Cases"
    )
)

fig.update_layout(

    legend=dict(

        orientation="h",

        y=1.08,

        x=0.5,

        xanchor="center"

    )

)

fig.show()

### ⚖️ Key Insight — Criminal Case Profile of Winning Candidates

A substantial difference is visible between the two major parties in the proportion of winners with **no declared criminal cases**. Among AITC's 80 elected candidates, **57.5% have no declared criminal cases**, compared with only **26.0% of BJP's 204 winners**. Consequently, **74.0% of BJP's elected candidates and 42.5% of AITC's elected candidates have at least one declared criminal case**.

The **100% clean-slate figures for INC, CPI(M), All India Secular Front and Aam Janata Unnayan Party should not be interpreted as strong party-level evidence**, because these parties have only 2, 1, 1 and 1 elected candidates respectively. A 100% figure based on one or two observations has very limited statistical significance.

### ⚠️ Political Context & Possible Implication

This finding becomes more noteworthy in the context of the **2026 election**, where AITC entered the election as the incumbent governing party and BJP emerged with the largest number of winners in this dataset. One possible line of investigation is whether BJP's candidate selection was associated with a **higher prevalence of candidates with declared criminal cases**, or whether differences arise from the political environment, constituency characteristics, opposition dynamics, or the types of cases involved.

However, the chart **cannot establish why these differences exist**. In particular, it would be inappropriate to conclude that BJP candidates are inherently more "prone" to criminality based solely on declared cases. A declared case is also **not equivalent to a conviction or proof of criminal conduct**.

### 📌 Political Consulting Takeaway

The **74.0% figure for BJP is the strongest signal in this chart and warrants further investigation**. If this pattern remains consistent across previous elections and is also associated with serious or multiple cases, it could represent a significant **candidate-selection and political-risk issue**. Conversely, if the cases are predominantly minor, politically motivated, or concentrated in particular constituencies, the interpretation would be very different.

A useful next step would therefore be to examine **case severity, number of cases per candidate, conviction status, constituency-wise distribution, and historical trends across previous West Bengal elections** rather than treating all criminal cases as equivalent.

In [ ]:
# @title 7. Candidates debt ratio
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import numpy as np
import pandas as pd

# ==========================================================
# Debt Ratio
# ==========================================================

df["debt_ratio"] = np.where(
    df["Assets_Amount"] > 0,
    df["Liabilities_Amount"] / df["Assets_Amount"],
    np.nan
)

# ==========================================================
# Categories
# ==========================================================

df["Debt Category"] = pd.cut(
    df["debt_ratio"],
    bins=[-0.01,0.25,0.50,1.00,np.inf],
    labels=[
        "Very Low Debt (<25%)",
        "Low Debt (25–50%)",
        "Moderate Debt (50–100%)",
        "Liabilities Exceed Assets (>100%)"
    ],
    right=False
)

debt_distribution = (
    df["Debt Category"]
      .value_counts()
      .sort_index()
      .reset_index(name="Number of Winners")
)

# ==========================================================
# Top 5 Candidates with Debt Ratio > 1
# ==========================================================

high_debt = (df[df["debt_ratio"]>=1]
    .sort_values("debt_ratio", ascending=False)
    .head(5)
    [["Candidate", "Constituency", "Party", "debt_ratio"]]
)

high_debt["debt_ratio"] = high_debt["debt_ratio"].round(2)

# ==========================================================
# Figure
# ==========================================================

fig = make_subplots(
    rows=1,
    cols=2,
    column_widths=[0.65,0.35],
    specs=[[{"type":"bar"},{"type":"table"}]]
)

# ----------------------------------------------------------
# Bar Chart
# ----------------------------------------------------------

colors = [
    "#2E7D32",
    "#81C784",
    "#F9A825",
    "#C62828"
]

fig.add_trace(

    go.Bar(

        x=debt_distribution["Debt Category"],
        y=debt_distribution["Number of Winners"],

        marker_color=colors,

        text=debt_distribution["Number of Winners"],

        textposition="outside",

        showlegend=False

    ),

    row=1,
    col=1

)
fig.add_annotation(
    x=0.94,
    y=1.08,
    xref="paper",
    yref="paper",
    text="<b>Candidates with Debt Ratio > 1</b>",
    showarrow=False,
    font=dict(size=16)
)
# ----------------------------------------------------------
# Table
# ----------------------------------------------------------

fig.add_trace(

    go.Table(

        header=dict(

            values=[
                "<b>Candidate</b>",
                "<b>Constituency</b>",
                "<b>Party</b>",
                "<b>Debt Ratio</b>"
            ],

            fill_color="#263238",

            font=dict(
                color="white",
                size=13
            ),

            align="left"

        ),

        cells=dict(

            values=[
                high_debt["Candidate"],
                high_debt["Constituency"],
                high_debt["Party"],
                high_debt["debt_ratio"]
            ],

            align="left",

            height=30

        )

    ),

    row=1,
    col=2

)

# ==========================================================
# Layout
# ==========================================================

fig.update_layout(

    title=dict(
        text="<b>Debt Burden of Winning Candidates</b>",
        x=0.5,
        font=dict(size=24)
    ),

    paper_bgcolor="white",
    plot_bgcolor="white",

    height=650,
    width=1350,

    margin=dict(
        l=40,
        r=40,
        t=80,
        b=40
    )

)

fig.update_xaxes(title="Debt Category", row=1, col=1)
fig.update_yaxes(title="Number of Winning Candidates", row=1, col=1)

fig.show()

/tmp/ipykernel_610/1703267658.py:11: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_610/1703267658.py:21: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



### 💳 Key Insight — Debt Burden of Winning Candidates

The financial profile is overwhelmingly low-debt: **239 of 289 candidates (82.7%)** have a debt ratio below 0.25, while only **4 candidates (1.4%)** have liabilities exceeding their declared assets.

### ⚠️ Notable Observation

The most interesting finding is that **3 of the 4 candidates with liabilities exceeding assets are from BJP** — Labsen Baskey (1.63), Karfa Soumen (1.37), and Satyendra Nath Roy (1.33), while Ahmed Javed Khan (AITC) records 1.11. This creates a potentially interesting **party-level financial-risk pattern**, although the sample is too small to conclude that BJP candidates generally carry higher debt.

These cases are worth examining further through the **original affidavits and liability composition**. A debt ratio above 1 is unusual, but by itself it is **not evidence of wrongdoing or financial misconduct**.

In [ ]:
df.groupby("Party")["debt_ratio"].mean()

,debt_ratio
Party,
AITC,0.086922
Aam Janata Unnayan party,0.000000
All India Secular Front,0.000000
BJP,0.141822
CPI(M),0.000000
INC,0.526160


In [ ]:
# @title 8. Average debt ratio of political party
import plotly.express as px

# ----------------------------------------------------------
# Exclude Parties
# ----------------------------------------------------------

exclude_parties = [
    "Aam Janata Unnayan party",
    "All India Secular Front",
    "CPI(M)"
]

party_debt = (
    df[~df["Party"].isin(exclude_parties)]
      .groupby("Party", as_index=False)["debt_ratio"]
      .mean()
      .round(2)
      .sort_values("debt_ratio", ascending=True)
)

# ----------------------------------------------------------
# Visualization
# ----------------------------------------------------------

fig = px.bar(
    party_debt,
    x="debt_ratio",
    y="Party",
    orientation="h",
    text="debt_ratio",
    color="debt_ratio",
    color_continuous_scale="Reds"
)

fig.update_traces(
    texttemplate="%{text:.2f}",
    textposition="outside",
    hovertemplate=
    "<b>%{y}</b><br>"
    "Average Debt Ratio: %{x:.2f}<extra></extra>"
)

fig.update_layout(

    title=dict(
        text="<b>Average Debt-to-Asset Ratio by Political Party</b>",
        x=0.5,
        font=dict(size=24)
    ),

    xaxis_title="Average Debt Ratio",

    yaxis_title="Political Party",

    yaxis=dict(
        categoryorder="total ascending"
    ),

    paper_bgcolor="white",
    plot_bgcolor="white",

    coloraxis_showscale=False,

    height=600,
    width=950,

    font=dict(
        family="Arial",
        size=13
    ),

    margin=dict(
        l=120,
        r=40,
        t=80,
        b=50
    )
)

fig.show()

In [ ]:
# @title 9. Lowest net worth(Poorest!!!)
import numpy as np
import plotly.graph_objects as go

# ==========================================================
# Bottom 20 Candidates by Net Worth
# (Displayed in descending order of Assets)
# ==========================================================

df_ = (
    df.sort_values("net_worth", ascending=True)
      .head(20)
      .sort_values("Assets_Amount", ascending=False)
      .copy()
)

# ==========================================================
# Convert to Lakhs (for Hover Information)
# ==========================================================

df_["Assets (₹ Lakh)"] = (df_["Assets_Amount"] / 1e5).round(2)
df_["Net Worth (₹ Lakh)"] = (df_["net_worth"] / 1e5).round(2)

# ==========================================================
# Signed Log Transformation
# ==========================================================

# Assets are assumed to be positive
df_["Assets_log"] = np.log10(df_["Assets_Amount"] + 1)

# Signed log transformation for Net Worth
df_["NetWorth_log"] = (
    np.sign(df_["net_worth"]) *
    np.log10(np.abs(df_["net_worth"]) + 1)
)

# ==========================================================
# Create Figure
# ==========================================================

fig = go.Figure()

# ----------------------------------------------------------
# Assets
# ----------------------------------------------------------

fig.add_trace(
    go.Bar(
        x=df_["Candidate"],
        y=df_["Assets_log"],
        name="Assets",
        marker_color="#1f77b4",

        customdata=np.stack(
            (
                df_["Assets (₹ Lakh)"],
                df_["Net Worth (₹ Lakh)"],
                df_["Party"],
                df_["Constituency"]
            ),
            axis=-1
        ),

        hovertemplate=
        "<b>%{x}</b><br>"
        "Party: %{customdata[2]}<br>"
        "Constituency: %{customdata[3]}<br><br>"
        "<b>Assets:</b> ₹ %{customdata[0]:,.2f} Lakh<br>"
        "<b>Net Worth:</b> ₹ %{customdata[1]:,.2f} Lakh"
        "<extra></extra>"
    )
)

# ----------------------------------------------------------
# Net Worth
# ----------------------------------------------------------

fig.add_trace(
    go.Bar(
        x=df_["Candidate"],
        y=df_["NetWorth_log"],
        name="Net Worth",
        marker_color="#2ca02c",

        customdata=np.stack(
            (
                df_["Assets (₹ Lakh)"],
                df_["Net Worth (₹ Lakh)"],
                df_["Party"],
                df_["Constituency"]
            ),
            axis=-1
        ),

        hovertemplate=
        "<b>%{x}</b><br>"
        "Party: %{customdata[2]}<br>"
        "Constituency: %{customdata[3]}<br><br>"
        "<b>Assets:</b> ₹ %{customdata[0]:,.2f} Lakh<br>"
        "<b>Net Worth:</b> ₹ %{customdata[1]:,.2f} Lakh"
        "<extra></extra>"
    )
)

# ==========================================================
# Layout
# ==========================================================

fig.update_layout(

    barmode="group",

    title=dict(
        text="<b>Assets vs Net Worth of the 20 Lowest Net Worth Winning Candidates</b>",
        x=0.5,
        font=dict(size=24)
    ),

    xaxis_title="Candidate",

    yaxis_title="Signed Log₁₀(Value)",

    paper_bgcolor="white",
    plot_bgcolor="white",

    height=700,
    width=1450,

    font=dict(
        family="Arial",
        size=13
    ),

    legend=dict(
        orientation="h",
        y=1.08,
        x=0.5,
        xanchor="center"
    ),

    margin=dict(
        l=80,
        r=40,
        t=90,
        b=200
    )
)

fig.update_xaxes(
    tickangle=-60
)

fig.show()

In [ ]:
summary = (
    df.groupby(["Education_level", "Party"])["Criminal Cases"]
      .agg(
          Candidates="count",
          Average_Cases="mean",
          Total_Cases="sum"
      )
      .round({"Average_Cases": 2})
      .reset_index()
)

In [ ]:
# @title 10. Criminal case profile by party and education level
import plotly.graph_objects as go

# ==========================================================
# Prepare Data
# ==========================================================

summary = (
    df.groupby(["Education_level", "Party"])["Criminal Cases"]
      .agg(
          Candidates="count",
          Average_Cases="mean",
          Total_Cases="sum"
      )
      .round({"Average_Cases": 2})
      .reset_index()
)

# ==========================================================
# Pivot Tables
# ==========================================================

avg_cases = summary.pivot(
    index="Education_level",
    columns="Party",
    values="Average_Cases"
)

candidate_count = summary.pivot(
    index="Education_level",
    columns="Party",
    values="Candidates"
)

total_cases = summary.pivot(
    index="Education_level",
    columns="Party",
    values="Total_Cases"
)

# ==========================================================
# Annotation Text
# ==========================================================

annotation_text = []

for i in range(len(avg_cases.index)):

    row = []

    for j in range(len(avg_cases.columns)):

        c = candidate_count.iloc[i, j]
        t = total_cases.iloc[i, j]

        if pd.isna(c):
            row.append("")
        else:
            row.append(f"{int(c)} | {int(t)}")

    annotation_text.append(row)

# ==========================================================
# Heatmap
# ==========================================================

fig = go.Figure(

    data=go.Heatmap(

        z=avg_cases.values,

        x=avg_cases.columns,

        y=avg_cases.index,

        colorscale="Reds",

        colorbar=dict(
            title="Average Criminal Cases"
        ),

        text=annotation_text,

        texttemplate="%{text}",

        textfont=dict(
            color="black",
            size=12
        ),

        hovertemplate=
        "<b>Party:</b> %{x}<br>" +
        "<b>Education:</b> %{y}<br>" +
        "<b>Average Criminal Cases:</b> %{z:.2f}<br>" +
        "<b>Candidates | Total Cases:</b> %{text}" +
        "<extra></extra>"

    )

)

# ==========================================================
# Layout
# ==========================================================

fig.update_layout(

    title=dict(
        text="<b>Criminal Case Profile by Party and Education Level</b>",
        x=0.5,
        font=dict(size=24)
    ),

    xaxis_title="Political Party",

    yaxis_title="Education Level",

    paper_bgcolor="white",

    plot_bgcolor="white",

    width=1100,

    height=450,

    font=dict(
        family="Arial",
        size=13
    )
)

fig.show()

In [ ]:
# @title 11. Relation between wealth and criminal cases
import plotly.graph_objects as go
from scipy.stats import spearmanr

# ==========================================================
# Calculate Spearman Correlation
# ==========================================================

corr, p_value = spearmanr(
    df["net_worth"],
    df["Criminal Cases"]
)

fig = go.Figure(go.Indicator(

    mode="gauge+number",

    value=corr,

    number={
        "valueformat": ".3f",
        "font": {"size": 36}
    },

    title={
        "text": "<b>Spearman Correlation (ρ)</b><br><span style='font-size:14px'>Net Worth vs Criminal Cases</span>"
    },

    gauge={

        "shape": "angular",

        "axis": {
            "range": [-1, 1],
            "tickvals": [-1, -0.5, 0, 0.5, 1],
            "ticktext": [
                "-1",
                "-0.5",
                "0",
                "0.5",
                "1"
            ]
        },

        # Hide the blue filled bar
        "bar": {
            "color": "rgba(0,0,0,0)",
            "thickness": 0
        },

        "steps": [
            {"range": [-1, -0.7], "color": "#d73027"},
            {"range": [-0.7, -0.3], "color": "#fc8d59"},
            {"range": [-0.3,  0.3], "color": "#fee08b"},
            {"range": [ 0.3,  0.7], "color": "#91bfdb"},
            {"range": [ 0.7,  1.0], "color": "#4575b4"}
        ],

        # Acts like a needle
        "threshold": {
            "line": {
                "color": "black",
                "width": 6
            },
            "thickness": 1,
            "value": corr
        }

    }

))

fig.update_layout(

    title=dict(
        text="<b>Relationship Between Net Worth and Criminal Cases</b>",
        x=0.5,
        font=dict(size=22)
    ),

    height=500,
    width=750,

    paper_bgcolor="white",

    annotations=[
        dict(
            x=0.5,
            y=0.25,
            xref="paper",
            yref="paper",
            showarrow=False,
            text=f"<b>P-value:</b> {p_value:.4f}",
            font=dict(size=15)
        )
    ]
)

fig.show()

In [ ]:
# @title 12. Criminal Case Profile Across Educational Qualifications of Winning Candidates
import plotly.graph_objects as go
import pandas as pd

# ==========================================================
# Education Order
# ==========================================================

education_order = [
    "Illiterate",
    "Literate",
    "5th Pass",
    "8th Pass",
    "10th Pass",
    "12th Pass",
    "Graduate",
    "Post Graduate",
    "Doctorate"
]

# ==========================================================
# Summary Statistics
# ==========================================================

education_summary = (
    df.groupby("Education")["Criminal Cases"]
      .agg(
          Total_Criminal_Cases="sum",
          Average_Criminal_Cases="mean",
          Candidates="count"
      )
      .round(2)
      .reset_index()
)

education_summary["Education"] = pd.Categorical(
    education_summary["Education"],
    categories=education_order,
    ordered=True
)

education_summary = education_summary.sort_values("Education")

# ==========================================================
# Create Figure
# ==========================================================

fig = go.Figure()

# ==========================================================
# Total Criminal Cases (Bars)
# ==========================================================

fig.add_trace(

    go.Bar(

        x=education_summary["Education"],

        y=education_summary["Total_Criminal_Cases"],

        name="Total Criminal Cases",

        marker=dict(

            color="#1F1F1F",

            line=dict(
                color="#000000",
                width=1.2
            )

        ),

        text=education_summary["Total_Criminal_Cases"],

        textposition="outside",

        customdata=education_summary[["Candidates"]],

        hovertemplate=
        "<b>%{x}</b><br><br>"
        "Total Criminal Cases: <b>%{y}</b><br>"
        "Winning Candidates: <b>%{customdata[0]}</b>"
        "<extra></extra>"

    )

)

# ==========================================================
# Average Criminal Cases (Line)
# ==========================================================

fig.add_trace(

    go.Scatter(

        x=education_summary["Education"],

        y=education_summary["Average_Criminal_Cases"],

        mode="lines+markers+text",

        name="Average Criminal Cases",

        line=dict(
            color="#7A0000",
            width=4
        ),

        marker=dict(
            size=11,
            color="white",
            line=dict(
                color="#7A0000",
                width=3
            )
        ),

        text=education_summary["Average_Criminal_Cases"],

        texttemplate="%{text:.2f}",

        textposition="top center",

        customdata=education_summary[["Candidates"]],

        hovertemplate=
        "<b>%{x}</b><br><br>"
        "Average Criminal Cases: <b>%{y:.2f}</b><br>"
        "Winning Candidates: <b>%{customdata[0]}</b>"
        "<extra></extra>",

        yaxis="y2"

    )

)

# ==========================================================
# Layout
# ==========================================================

fig.update_layout(

    title=dict(

        text="<b>Criminal Case Profile Across Educational Qualifications of Winning Candidates</b>",

        x=0.5,

        font=dict(size=24)

    ),

    xaxis=dict(

        title="<b>Educational Qualification</b>",

        showgrid=False,

        showline=True,

        linecolor="#444444",

        linewidth=1.5

    ),

    yaxis=dict(

        title="<b>Total Criminal Cases</b>",

        showgrid=True,

        gridcolor="#E8E8E8",

        zeroline=False

    ),

    yaxis2=dict(

        title="<b>Average Criminal Cases</b>",

        overlaying="y",

        side="right",

        showgrid=False,

        zeroline=False

    ),

    legend=dict(

        orientation="h",

        y=1.08,

        x=0.5,

        xanchor="center"

    ),

    paper_bgcolor="white",

    plot_bgcolor="white",

    font=dict(

        family="Arial",

        size=13,

        color="#222222"

    ),

    height=650,

    width=1250,

    margin=dict(

        l=70,

        r=70,

        t=90,

        b=70

    )

)

fig.show()

In [ ]:
# @title 13. Which party has highest total criminal case?
import plotly.graph_objects as go
import pandas as pd

# ==========================================================
# Party-wise Criminal Case Summary
# ==========================================================

party_crime = (
    df.groupby("Party")["Criminal Cases"]
      .agg(
          Total_Criminal_Cases="sum",
          Average_Criminal_Cases="mean",
          Winners="count"
      )
      .round(2)
      .sort_values("Total_Criminal_Cases", ascending=False)
      .reset_index()
)

# ==========================================================
# Party-wise Candidates with Declared Criminal Cases
# ==========================================================

criminal_candidates = (
    df[df["Criminal Cases"] > 0]["Party"]
      .value_counts()
)

# ----------------------------------------------------------
# Create Note
# ----------------------------------------------------------


note = (
    "<span style='font-size:15px;'>"
    "<b>Party-wise no. of Candidates with<br>"
    "Declared Criminal Cases from total</b></span><br><br>"
)

for party, count in criminal_candidates.items():
    note += (
        f"<span style='color:#8B0000;'>●</span> "
        f"<b>{party}</b> : "
        f"<span style='color:#1F1F1F;'><b>{count}</b></span><br>"
    )
# ==========================================================
# Party Colors
# ==========================================================

party_colors = {
    "BJP": "#FF9933",                         # Saffron
    "AITC": "#00A651",                        # Green
    "INC": "#19A0FF",                         # Blue
    "CPI(M)": "#D50000",                      # Red
    "Aam Janata Unnayan Party": "#6A1B9A",
    "All India Secular Front": "#00897B",
    "Others": "#666666"
}

# ==========================================================
# Create Color-coded Note
# ==========================================================

note = (
    "<span style='font-size:15px;'>"
    "<b>Party-wise no. of Candidates with<br>"
    "Declared Criminal Cases from total.</b></span><br><br>"
)

for party, count in criminal_candidates.items():

    color = party_colors.get(party, "#666666")   # Default grey

    note += (
        f"<span style='color:{color};'><b>● {party}</b></span>"
        f" : <span style='color:#1F1F1F;'><b>{count}</b></span><br>"
    )
# ==========================================================
# Figure
# ==========================================================

fig = go.Figure()

# ----------------------------------------------------------
# Bars : Total Criminal Cases
# ----------------------------------------------------------

fig.add_trace(

    go.Bar(

        x=party_crime["Party"],

        y=party_crime["Total_Criminal_Cases"],

        name="Total Criminal Cases",

        marker=dict(
            color="#1F1F1F",
            line=dict(
                color="black",
                width=1.2
            )
        ),

        text=party_crime["Total_Criminal_Cases"],

        textposition="outside",

        customdata=party_crime[
            ["Average_Criminal_Cases", "Winners"]
        ],

        hovertemplate=
        "<b>%{x}</b><br><br>"
        "Total Criminal Cases : <b>%{y}</b><br>"
        "Average Criminal Cases : <b>%{customdata[0]:.2f}</b><br>"
        "Winning Candidates : <b>%{customdata[1]}</b>"
        "<extra></extra>"

    )

)

# ----------------------------------------------------------
# Line : Average Criminal Cases
# ----------------------------------------------------------

fig.add_trace(

    go.Scatter(

        x=party_crime["Party"],

        y=party_crime["Average_Criminal_Cases"],

        mode="lines+markers+text",

        name="Average Criminal Cases",

        yaxis="y2",

        line=dict(
            color="#7A0000",
            width=4
        ),

        marker=dict(
            size=11,
            color="white",
            line=dict(
                color="#7A0000",
                width=3
            )
        ),

        text=party_crime["Average_Criminal_Cases"],

        texttemplate="%{text:.2f}",

        textposition="top center",

        hovertemplate=
        "<b>%{x}</b><br>"
        "Average Criminal Cases : <b>%{y:.2f}</b>"
        "<extra></extra>"

    )

)

# ==========================================================
# Side Note
# ==========================================================

fig.add_annotation(

    x=1.02,
    y=0.98,

    xref="paper",
    yref="paper",

    text=note,

    showarrow=False,

    align="left",

    bordercolor="#444444",
    borderwidth=1.2,

    borderpad=10,

    bgcolor="#F8F8F8",

    font=dict(
        family="Arial",
        size=12,
        color="#222222"
    )

)

# ==========================================================
# Layout
# ==========================================================

fig.update_layout(

    title=dict(

        text="<b>Party-wise Criminal Case Profile of Winning Candidates</b>",

        x=0.5,

        font=dict(size=24)

    ),

    xaxis=dict(

        title="<b>Political Party</b>",

        tickangle=-30,

        showgrid=False,

        showline=True,

        linecolor="#444444"

    ),

    yaxis=dict(

        title="<b>Total Criminal Cases</b>",

        gridcolor="#ECECEC",

        zeroline=False

    ),

    yaxis2=dict(

        title="<b>Average Criminal Cases</b>",

        overlaying="y",

        side="right",

        showgrid=False,

        zeroline=False

    ),

    legend=dict(

        orientation="h",

        x=0.5,

        xanchor="center",

        y=1.08

    ),

    paper_bgcolor="white",

    plot_bgcolor="white",

    font=dict(
        family="Arial",
        size=13,
        color="#222222"
    ),

    margin=dict(
        l=70,
        r=320,
        t=90,
        b=80
    ),

    width=1450,

    height=700

)

fig.show()

In [ ]:
# @title 14. Is their a corelation between education and assets?
import plotly.graph_objects as go
import pandas as pd

# ==========================================================
# Average Assets + Candidate Count by Education
# ==========================================================

education_assets = (
    df.groupby("Education")
      .agg(
          Mean_Assets=("Assets_Amount", "mean"),
          Candidates=("Candidate", "count")
      )
      .sort_values("Mean_Assets", ascending=False)
      .reset_index()
)

# Convert mean assets to Crores
education_assets["Mean_Assets_Cr"] = (
    education_assets["Mean_Assets"] / 1e7
).round(2)

# ==========================================================
# Indian Financial Theme
# ==========================================================

DARK_GREEN = "#0B3D2E"
GREEN = "#146B4B"
GOLD = "#C9A227"
OFF_WHITE = "#FCFBF5"
DARK_TEXT = "#1E2A24"
GRID = "#E4E0D2"

# ==========================================================
# Figure
# ==========================================================

fig = go.Figure()

# ==========================================================
# BAR — Mean Assets
# ==========================================================

fig.add_trace(
    go.Bar(
        y=education_assets["Education"],
        x=education_assets["Mean_Assets"],
        orientation="h",

        name="Mean Assets",

        marker=dict(
            color=GREEN,
            line=dict(
                color=DARK_GREEN,
                width=1.5
            )
        ),

        # Candidate count INSIDE the bar
        text=education_assets["Candidates"].apply(
            lambda x: f"{x} candidates"
        ),

        textposition="inside",

        insidetextanchor="middle",

        textfont=dict(
            color="white",
            size=13
        ),

        customdata=education_assets[
            [
                "Mean_Assets_Cr",
                "Candidates"
            ]
        ],

        hovertemplate=
        "<b>%{y}</b><br><br>"
        "Mean Assets: "
        "<b>₹ %{customdata[0]:,.2f} Cr</b><br>"
        "Candidates: "
        "<b>%{customdata[1]}</b>"
        "<extra></extra>"
    )
)

# ==========================================================
# MEAN WEALTH LABEL — AT END OF BAR
# ==========================================================

fig.add_trace(
    go.Scatter(

        y=education_assets["Education"],

        x=education_assets["Mean_Assets"],

        mode="text",

        text=education_assets["Mean_Assets_Cr"].apply(
            lambda x: f"₹ {x:,.2f} Cr"
        ),

        textposition="middle right",

        textfont=dict(
            family="Arial",
            size=13,
            color=DARK_GREEN
        ),

        hoverinfo="skip",

        showlegend=False

    )
)

# ==========================================================
# Layout
# ==========================================================

fig.update_layout(

    title=dict(
        text=(
            "<span style='color:#0B3D2E'>"
            "<b>Average Declared Assets by Educational Qualification</b>"
            "</span>"
            "<br>"
            "<span style='font-size:14px;color:#6B6B5E'>"
            "Winning Candidates • Indian Election Wealth Profile"
            "</span>"
        ),
        x=0.5,
        font=dict(
            size=25,
            color=DARK_TEXT
        )
    ),

    xaxis=dict(

        title="<b>Average Declared Assets (₹)</b>",

        type="log",

        gridcolor=GRID,

        zeroline=False,

        tickprefix="₹ ",

        showline=True,

        linecolor="#B8B4A6"

    ),

    yaxis=dict(

        title="<b>Educational Qualification</b>",

        categoryorder="array",

        categoryarray=education_assets["Education"].tolist(),

        showgrid=False,

        tickfont=dict(
            size=13,
            color=DARK_TEXT
        )

    ),

    paper_bgcolor=OFF_WHITE,

    plot_bgcolor=OFF_WHITE,

    font=dict(
        family="Arial",
        size=13,
        color=DARK_TEXT
    ),

    width=1250,

    height=680,

    margin=dict(
        l=150,
        r=180,
        t=120,
        b=90
    ),

    showlegend=False
)

# ==========================================================
# Gold Accent
# ==========================================================

fig.add_shape(
    type="line",

    x0=0,
    x1=1,

    y0=1.01,
    y1=1.01,

    xref="paper",
    yref="paper",

    line=dict(
        color=GOLD,
        width=3
    )
)

# ==========================================================
# Footer
# ==========================================================

fig.add_annotation(

    x=0.5,
    y=-0.14,

    xref="paper",
    yref="paper",

    text=(
        "<span style='color:#C9A227'>₹</span> "
        "<b>Bar length = mean declared assets</b>"
        " &nbsp;•&nbsp; "
        "<b>Number inside bar = candidates</b>"
        " &nbsp;•&nbsp; "
        "Values shown in Crores"
    ),

    showarrow=False,

    font=dict(
        size=11,
        color="#6B6B5E"
    )
)

fig.show()

In [ ]:
# @title
education_rank = {
    "Illiterate": 0,
    "Literate": 1,
    "5th Pass": 2,
    "8th Pass": 3,
    "10th Pass": 4,
    "12th Pass": 5,
    "Graduate": 6,
    "Post Graduate": 7,
    "Doctorate": 8
}

df["Education_Rank"] = df["Education"].map(education_rank)

import plotly.graph_objects as go
import numpy as np
from scipy.stats import spearmanr

# ==========================================================
# EDUCATION ORDER
# ==========================================================

education_order = [
    "Illiterate",
    "Literate",
    "5th Pass",
    "8th Pass",
    "10th Pass",
    "12th Pass",
    "Graduate",
    "Post Graduate",
    "Doctorate"
]

education_rank = {
    education: i
    for i, education in enumerate(education_order)
}

# ==========================================================
# PREPARE DATA
# ==========================================================

analysis_df = df[
    [
        "Education",
        "Education_Rank",
        "Assets_Amount",
        "Candidate",
        "Party",
        "Constituency"
    ]
].copy()

analysis_df = analysis_df.dropna(
    subset=["Education_Rank", "Assets_Amount"]
)

analysis_df = analysis_df[
    analysis_df["Assets_Amount"] > 0
]

# ==========================================================
# SPEARMAN CORRELATION
# ==========================================================

corr, p_value = spearmanr(
    analysis_df["Education_Rank"],
    analysis_df["Assets_Amount"]
)

# ==========================================================
# JITTER EDUCATION RANK
# ==========================================================

np.random.seed(42)

analysis_df["Education_Jitter"] = (
    analysis_df["Education_Rank"]
    + np.random.uniform(
        -0.16,
        0.16,
        len(analysis_df)
    )
)

# ==========================================================
# INTERPRETATION
# ==========================================================

if abs(corr) < 0.10:
    relationship = "Very Weak / No Monotonic Relationship"

elif abs(corr) < 0.30:
    relationship = "Weak Relationship"

elif abs(corr) < 0.50:
    relationship = "Moderate Relationship"

elif abs(corr) < 0.70:
    relationship = "Strong Relationship"

else:
    relationship = "Very Strong Relationship"


significance = (
    "Statistically Significant"
    if p_value < 0.05
    else "Not Statistically Significant"
)

# ==========================================================
# FIGURE
# ==========================================================

fig = go.Figure()

# ==========================================================
# CANDIDATE POINTS
# ==========================================================

fig.add_trace(

    go.Scatter(

        x=analysis_df["Education_Jitter"],

        y=analysis_df["Assets_Amount"],

        mode="markers",

        name="Winning Candidates",

        marker=dict(

            size=9,

            color="#263238",

            opacity=0.65,

            line=dict(
                color="white",
                width=1
            )

        ),

        customdata=analysis_df[
            [
                "Education",
                "Candidate",
                "Party",
                "Constituency"
            ]
        ],

        hovertemplate=

        "<b>%{customdata[1]}</b><br>"
        "Education: %{customdata[0]}<br>"
        "Party: %{customdata[2]}<br>"
        "Constituency: %{customdata[3]}<br>"
        "Assets: ₹ %{y:,.0f}"
        "<extra></extra>"

    )

)

# ==========================================================
# MEAN ASSETS BY EDUCATION
# ==========================================================

mean_assets = (
    analysis_df
    .groupby("Education_Rank")["Assets_Amount"]
    .mean()
    .reindex(range(len(education_order)))
)

# ==========================================================
# MEAN TREND LINE
# ==========================================================

fig.add_trace(

    go.Scatter(

        x=list(range(len(education_order))),

        y=mean_assets.values,

        mode="lines+markers",

        name="Mean Assets",

        line=dict(
            color="#8B0000",
            width=4
        ),

        marker=dict(

            size=11,

            color="#8B0000",

            line=dict(
                color="white",
                width=2
            )

        ),

        hovertemplate=

        "<b>%{text}</b><br>"
        "Mean Assets: ₹ %{y:,.0f}"
        "<extra></extra>",

        text=education_order

    )

)

# ==========================================================
# LAYOUT
# ==========================================================

fig.update_layout(

    title=dict(

        text=(
            "<b>Educational Attainment vs Candidate Wealth</b>"
            "<br>"
            "<span style='font-size:15px;color:#555'>"
            "Distribution of Assets Across Educational Qualifications"
            "</span>"
        ),

        x=0.5,

        font=dict(
            size=25,
            color="#1F1F1F"
        )

    ),

    xaxis=dict(

        title="<b>Educational Qualification</b>",

        tickmode="array",

        tickvals=list(range(len(education_order))),

        ticktext=education_order,

        tickangle=-35,

        range=[
            -0.5,
            len(education_order) - 0.5
        ],

        showgrid=False,

        showline=True,

        linecolor="#555555",

        linewidth=1.5

    ),

    yaxis=dict(

        title="<b>Assets (₹) — Log Scale</b>",

        type="log",

        gridcolor="#E5E5E5",

        gridwidth=1,

        zeroline=False

    ),

    paper_bgcolor="white",

    plot_bgcolor="#FAFAFA",

    font=dict(

        family="Arial",

        size=13,

        color="#222222"

    ),

    width=1350,

    height=720,

    margin=dict(

        l=90,

        r=300,

        t=120,

        b=130

    ),

    legend=dict(

        orientation="h",

        x=0.5,

        xanchor="center",

        y=1.02

    )

)

# ==========================================================
# STATISTICAL INSIGHT PANEL
# ==========================================================

insight_text = (

    "<b>RELATIONSHIP SUMMARY</b><br><br>"

    f"<span style='font-size:22px;color:#8B0000'>"
    f"<b>ρ = {corr:.3f}</b>"
    f"</span><br>"

    f"<b>Spearman Correlation</b><br><br>"

    f"<b>P-value:</b> {p_value:.4f}<br><br>"

    f"<b>{relationship}</b><br>"

    f"<span style='color:#555'>"
    f"{significance}"
    f"</span><br><br>"

    "● <b>Each dot</b> = winning candidate<br>"
    "● <b>Dark red line</b> = mean assets<br>"
    "● <b>Log scale</b> = accommodates large wealth differences"

)

fig.add_annotation(

    x=1.30,

    y=0.05,

    xref="paper",

    yref="paper",

    text=insight_text,

    showarrow=False,

    align="left",

    valign="top",

    bordercolor="#8B0000",

    borderwidth=1.5,

    borderpad=14,

    bgcolor="#F7F7F7",

    font=dict(

        family="Arial",

        size=13,

        color="#222222"

    )

)

# ==========================================================
# SHOW
# ==========================================================

fig.show()

# ==========================================================
# PRINT STATISTICS
# ==========================================================

print(
    "Spearman Correlation:",
    round(corr, 3)
)

print(
    "P-value:",
    round(p_value, 4)
)

/tmp/ipykernel_610/2621515257.py:14: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Spearman Correlation: 0.097
P-value: 0.106


In [ ]:
df_top5 = df[df["Education_level"]=="Higher Education"]
df_top5 = df_top5.sort_values(by="debt_ratio",ascending=True)[:100]
df_top5 = df_top5.sort_values(by="Criminal Cases",ascending=True)[:84]
df_top5 = df_top5.sort_values(by="Assets_Amount",ascending=False).head(10)
df_top5

,Sno,Candidate,Constituency,Party,Criminal Cases,Education,Assets_Amount,Liabilities_Amount,Education_level,net_worth,debt_ratio,Debt Category,Education_Rank
206,207,Ritesh Tiwari,KASHIPUR-BELGACHHIA,BJP,4,Graduate,278898708,0,Higher Education,278898708,0.000000,Very Low Debt (<25%),6.0
103,104,Firhad Hakim,KOLKATA PORT,AITC,2,Graduate,225753159,13679175,Higher Education,212073984,0.060594,Very Low Debt (<25%),6.0
180,181,Partha Ghosh,RANIGANJ,BJP,0,Post Graduate,177664006,3513888,Higher Education,174150118,0.019778,Very Low Debt (<25%),7.0
235,236,Sharadwat Mukherjee,BIDHANNAGAR,BJP,0,Post Graduate,157226453,4830854,Higher Education,152395599,0.030725,Very Low Debt (<25%),7.0
220,221,Sandipan Saha,ENTALLY,AITC,0,Graduate,119379822,3000000,Higher Education,116379822,0.025130,Very Low Debt (<25%),6.0
242,243,Sobhandeb Chattopadhyay,BALLYGUNGE,AITC,0,Graduate,102223379,0,Higher Education,102223379,0.000000,Very Low Debt (<25%),6.0
18,19,Amlan Bhaduri (Buro),ENGLISHBAZAR,BJP,1,Post Graduate,95139936,4040852,Higher Education,91099084,0.042473,Very Low Debt (<25%),7.0
209,210,Rupa Ganguly,SONARPUR DAKSHIN,BJP,5,Graduate,93954158,0,Higher Education,93954158,0.000000,Very Low Debt (<25%),6.0
95,96,Dr. Mosarraf Hossain,MURARAI,AITC,0,Graduate,88707097,703525,Higher Education,88003572,0.007931,Very Low Debt (<25%),6.0
192,193,Pulak Roy,ULUBERIA DAKSHIN,AITC,0,Graduate,86836465,4966980,Higher Education,81869485,0.057199,Very Low Debt (<25%),6.0


In [ ]:
df

,Sno,Candidate,Constituency,Party,Criminal Cases,Education,Assets_Amount,Liabilities_Amount,Education_level,net_worth,debt_ratio,Debt Category,Education_Rank
0,1,Abdul Aziz Doctor,LALGOLA,AITC,0,Post Graduate,62908242,3850000,Higher Education,59058242,0.061200,Very Low Debt (<25%),7.0
1,2,Abdul Khaleque Molla,METIABURUZ,AITC,2,10th Pass,17455505,0,Lower Education,17455505,0.000000,Very Low Debt (<25%),4.0
2,3,Abdul Matin Muhammad,HAROA,AITC,1,12th Pass,10014348,0,Lower Education,10014348,0.000000,Very Low Debt (<25%),5.0
3,4,Abdur Rahim Boxi,MALATIPUR,AITC,2,12th Pass,58283056,1049505,Lower Education,57233551,0.018007,Very Low Debt (<25%),5.0
4,5,Adhikari Dibyendu,EGRA,BJP,2,Graduate,48005224,5059168,Higher Education,42946056,0.105388,Very Low Debt (<25%),6.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
287,288,Tusar Kumar Majumdar,PANDUA,BJP,2,12th Pass,31568724,0,Lower Education,31568724,0.000000,Very Low Debt (<25%),5.0
288,289,Umesh Rai,HOWRAH UTTAR,BJP,19,Graduate,7617624,0,Higher Education,7617624,0.000000,Very Low Debt (<25%),6.0
289,290,Usha Rani Mondal,MINAKHAN (SC),AITC,0,8th Pass,23758335,1968000,Lower Education,21790335,0.082834,Very Low Debt (<25%),3.0
290,291,Utpal Brahmacharo,KALIAGANJ (SC),BJP,0,Graduate,4434501,0,Higher Education,4434501,0.000000,Very Low Debt (<25%),6.0
